oio/xcute/orchestrator.py

```diff
@@ -375,77 +371,154 @@ def process_reply(self, beanstalkd_job_id, encoded_reply):
+                if not beanstalkd:
                     continue
+                beanstalkd_workers[beanstalkd.addr] = beanstalkd
+            except Exception as exc:
+                self.logger.error('Fail to check beanstalkd: %s', exc)
+        return beanstalkd_workers
 
-                self.logger.info('Removed beanstalkd %s' % beanstalkd_addr)
-                del self.all_beanstalkd[beanstalkd_addr]
-
-            for beanstalkd_addr, beanstalkd \
-                    in all_beanstalkd_with_tubes.iteritems():
-                if beanstalkd_addr not in self.all_beanstalkd:
-                    self.logger.info('Found beanstalkd %s' % beanstalkd_addr)
-
-                self.all_beanstalkd[beanstalkd_addr] = beanstalkd + ({},)
-
-            sleep(5)
-
-        self.logger.info('Exited beanstalkd thread')
-
-    @staticmethod
-    def get_beanstalkd_tubes(beanstalkd_addr):
-        return Beanstalk.from_url('beanstalkd://' + beanstalkd_addr).tubes()
-
-    def get_loadbalanced_workers(self):
+    def _check_beanstalkd_worker(self, beanstalkd_info):
+        """
+        Check beanstalkd worker by looking at the score,
+        existing tubes and tube statistics.
         """
-            Yield senders following a loadbalancing strategy
+        beanstalkd_addr = 'beanstalk://' + beanstalkd_info['addr']
+        beanstalkd_score = beanstalkd_info['score']
+        if beanstalkd_score == 0:
+            self.logger.info(
+                'Ignore beanstalkd %s: score=0', beanstalkd_addr)
+            return None
+
+        beanstalkd = self.beanstalkd_workers.get(beanstalkd_addr)
+        if not beanstalkd:
+            beanstalkd = Beanstalk.from_url(beanstalkd_addr)
+            beanstalkd.addr = beanstalkd_addr
+            beanstalkd.use(self.beanstalkd_workers_tube)
+            beanstalkd.watch(self.beanstalkd_workers_tube)
+
+        beanstalkd_tubes = beanstalkd.tubes()
+        if self.beanstalkd_workers_tube not in beanstalkd_tubes:
+            self.logger.info(
+                'Ignore beanstalkd %s: '
+                'No worker has ever listened to the tube %s',
+                beanstalkd_addr, self.beanstalkd_workers_tube)
+            return None
+
+        current_stats = beanstalkd.stats_tube(
+            self.beanstalkd_workers_tube)
+        beanstalkd_jobs_ready = current_stats['current-jobs-ready']
+        if beanstalkd_jobs_ready > 0:
+            if current_stats['current-jobs-reserved'] <= 0:
+                self.logger.info(
+                    'Ignore beanstalkd %s: The worker doesn\'t process task',
+                    beanstalkd_addr)
+                return None
+
+            if beanstalkd_jobs_ready >= self.max_jobs_per_beanstalkd:
+                self.logger.info(
+                    'Ignore beanstalkd %s: The queue is full (%d jobs)',
+                    beanstalkd_addr, beanstalkd_jobs_ready)
+                return None
+
+        beanstalkd.is_broken = False
```

Comment:
That'd be cool to have a log in the case the beanstalkd was broken that says we're adding it back in the pool!